{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "# define a function that reads 8 lines at at time ones and Pcess it. Until the end of the fI\n",
    "def process_file(filename):\n",
    "    while True:\n",
    "        lines = [filename.readline() for _ in range(8)]\n",
    "        \n",
    "        if not any(lines):\n",
    "            break\n",
    "        #Pcess the lines\n",
    "        lines=''.join(lines).strip().replace('\\n','').replace(' ','').replace('t','u')\n",
    "        #print(lines)\n",
    "        #find the start codon and translate the sequence until the stop codon. Make tH recur\n",
    "        seq = []\n",
    "        def translate(lines, start_codon):\n",
    "            \n",
    "            for i in range(start_codon, len(lines)+1, 3):\n",
    "                codon = lines[i:i+3]\n",
    "                if codon in codon_table:\n",
    "                    if codon_table[codon] == 'Stop':\n",
    "                        seq.append(lines[start_codon:i])\n",
    "                        lines=lines[i+3:]\n",
    "                        start_codon = lines.find('aug') #start looking for the next start codon after the stop codon\n",
    "                        #print(start_codon)\n",
    "                        translate(lines, start_codon+3)\n",
    "                        break #\n",
    "        \n",
    "                    \n",
    "            #xxx.append(amino_acid)\n",
    "            return seq\n",
    "        \n",
    "        start_codon = lines.find('aug')     \n",
    "        seq.append(translate(lines, start_codon+3))\n",
    "        \n",
    "        for item in seq:\n",
    "            #take codons and translate them to amino acids\n",
    "            amino_acid = []\n",
    "            for k in range(0, len(item), 3):\n",
    "                try:\n",
    "                    codon = item[k:k+3]\n",
    "                    if codon in codon_table:\n",
    "                        amino_acid.append(codon_table[codon])\n",
    "                except:\n",
    "                    pass\n",
    "                \n",
    "            print(''.join(amino_acid))\n",
    "        print(\"---------------------\")\n",
    "            \n",
    "            \n",
    "            \n",
    "#dictionary of codons and their corresponding amino acids\n",
    "codon_table = {\n",
    "    'uuu': 'F', 'uuc': 'F', 'uua': 'L', 'uug': 'L',\n",
    "    'ucu': 'S', 'ucc': 'S', 'uca': 'S', 'ucg': 'S',\n",
    "    'uau': 'Y', 'uac': 'Y', 'uaa': 'Stop', 'uag': 'Stop',\n",
    "    'ugu': 'C', 'ugc': 'C', 'uga': 'Stop', 'ugg': 'W',\n",
    "    'cuu': 'L', 'cuc': 'L', 'cua': 'L', 'cug': 'L',\n",
    "    'ccu': 'P', 'ccc': 'P', 'cca': 'P', 'ccg': 'P',\n",
    "    'cau': 'H', 'cac': 'H', 'caa': 'Q', 'cag': 'Q',\n",
    "    'cgu': 'R', 'cgc': 'R', 'cga': 'R', 'cgg': 'R',\n",
    "    'auu': 'I', 'auc': 'I', 'aua': 'I', 'aug': 'M',\n",
    "    'acu': 'T', 'acc': 'T', 'aca': 'T', 'acg': 'T',\n",
    "    'aau': 'N', 'aac': 'N', 'aaa': 'L', 'aag': 'L',\n",
    "    'agu': 'S', 'agc': 'S', 'aga': 'R', 'agg': 'R',\n",
    "    'guu': 'V', 'guc': 'V', 'gua': 'V', 'gug': 'V',\n",
    "    'gcu': 'A', 'gcc': 'A', 'gca': 'A', 'gcg': 'A',\n",
    "    'gau': 'D', 'gac': 'D', 'gaa': 'E', 'gag': 'E',\n",
    "    'ggu': 'G', 'ggc': 'G', 'gga': 'G', 'ggg': 'G'\n",
    "}\n",
    "        \n",
    "\n",
    "\n",
    "#open the file and call the function\n",
    "filename='hexokinase.txt'\n",
    "with open(filename) as f:\n",
    "    process_file(f)\n",
    "        \n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
